File: 0039-combination-sum/0039-combination-sum.cpp
class Solution {
public:
    vector<int> path;
    vector<vector<int>> ans;
    void dfs(const vector<int>& candidates, int n, int start, int remaining){
        if (remaining == 0)
            ans.push_back(path);
        if (remaining < 0)
            return;
        for (int i=start; i<n; i++){
            int candidate = candidates[i];
            path.push_back(candidate);
            dfs(candidates,n,i,remaining-candidate);
            path.pop_back();
        }
    }
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        dfs(candidates,(int)candidates.size(),0,target);
        return ans;
    }
};

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        ans  = []
        path = []
        n    = len(candidates)
        def dfs(candidates: List[int], n: int, start: int, remaining: int) -> None:
            if remaining == 0:
                ans.append(path.copy())
                return
            if remaining < 0:
                return 
            for i in range(start,n):
                path.append(candidates[i])
                dfs(candidates,n,i,remaining-candidates[i])
                path.remove(candidates[i])
            return
        dfs(candidates,n,0,target)
        return ans